### PR TITLE
[ADD] feat(consumer): Consumer to clear 'filler' values like NA or N/A

### PR DESCRIPTION
**Pull request by Pourbaix**
A configurable consumer has been added that will check for values like NA or N/A in the item's metadata. It is only triggered when the item is not in workspace. In general, it will remove the 'filler' metadata values but if the field is in a form group, it will replace them with a placeholder value.

## Checklist

- [x] Check the code style using the command `mvn clean && mvn install` on local desktop
- [ ] Run unitest for `dspace-uclouvain` module

### CODE REVIEW

**Comment by Pourbaix** [comment]
I still have to move the changes in **`dspace-api/src/main/java/org/dspace/uclouvain/core/utils/ItemUtils.java`** to `uclouvain-base`.

File: dspace/config/modules/uclouvain.cfg [hunk 1]
```diff
@@ -34,6 +34,14 @@ event.consumer.authoritymetadataenhancer.filters = Item+All
 
 event.consumer.formfieldcleaner.class = org.dspace.uclouvain.consumer.FormFieldCleanerConsumer
 event.consumer.formfieldcleaner.filters = Item+All
+
+event.consumer.fillervalueremover.class = org.dspace.uclouvain.consumer.FillerValueRemoverConsumer
+event.consumer.fillervalueremover.filters = Item+Modify|Modify_Metadata
+event.consumer.fillervalueremover.fillers = NA
+event.consumer.fillervalueremover.fillers = na
+event.consumer.fillervalueremover.fillers = N/A
+event.consumer.fillervalueremover.fillers = n/a
```
Comment (zannkukai) [inline dspace/config/modules/uclouvain.cfg]: use a regexp to be more flexible --> `"\\s*(?i)(n/?a|not specified|no(t|n) applicable)\\s*"`

`\\s*` : n'importe quel nombre de d'espaces
`(?i)` : insensible à la case

File: dspace-api/src/main/java/org/dspace/uclouvain/consumer/FillerValueRemoverConsumer.java [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+     * 
+     * @param context The current DSpace context.
+     * @param event The event to handle in the consumer.
+     * @throws Exception
+     */
+    @Override
+    public void consume(Context context, Event event) throws Exception {
+        // Check that item can be processed (in archive).
+        Item item = (Item) event.getSubject(context);
+        if (item != null && workspaceItemService.findByItem(context, item) == null) {
```
Comment (zannkukai) [inline dspace-api/src/main/java/org/dspace/uclouvain/consumer/FillerValueRemoverConsumer.java]: use `!ItemUtils.isWorkspace(context, item)`

File: dspace-api/src/main/java/org/dspace/uclouvain/consumer/FillerValueRemoverConsumer.java [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+            DSpaceServicesFactory
+                .getInstance()
+                .getConfigurationService()
+                .getArrayProperty("event.consumer.fillervalueremover.fillers", new String[0])
+        );
+        workspaceItemService = ContentServiceFactory.getInstance().getWorkspaceItemService();
+        itemService = ContentServiceFactory.getInstance().getItemService();
+        submissionConfigService = SubmissionServiceFactory.getInstance().getSubmissionConfigService();
+        dciReader = new DCInputsReader();
+        itemToProcess = new HashSet<UUID>();
```
Comment (zannkukai) [inline dspace-api/src/main/java/org/dspace/uclouvain/consumer/FillerValueRemoverConsumer.java]: You can remove `UUID` --> `new HashSet<>()`

File: dspace-api/src/main/java/org/dspace/uclouvain/consumer/FillerValueRemoverConsumer.java [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+                }
+
+                // Metadata to remove.
+                List<MetadataValue> mvsToClear = new ArrayList<>();
+                // Metadata to replace with a placeholder.
+                List<MetadataValue> mvsToReplace = new ArrayList<>();
+
+                item.getMetadata().stream()
+                    .filter(metadata -> {
+                        String valueToCheck = metadata.getValue().trim();
```
Comment (zannkukai) [inline dspace-api/src/main/java/org/dspace/uclouvain/consumer/FillerValueRemoverConsumer.java]: could be remove, because regexp will manage this

File: dspace-api/src/main/java/org/dspace/uclouvain/consumer/FillerValueRemoverConsumer.java [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+                        return valueToCheck.isEmpty() || valuesToClear.contains(valueToCheck);
```
Comment (zannkukai) [inline dspace-api/src/main/java/org/dspace/uclouvain/consumer/FillerValueRemoverConsumer.java]: don't use "isEmpty" but "isBlank"

File: dspace-api/src/main/java/org/dspace/uclouvain/consumer/FillerValueRemoverConsumer.java [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+                    })
+                    .forEach(md -> {
+                        if (formGroupFields.contains(md.getMetadataField().toString('.'))) {
+                            mvsToReplace.add(md);
+                        } else {
+                            mvsToClear.add(md);
+                        }
+                    });
+
+                boolean metadataModified = false;
```
Comment (zannkukai) [inline dspace-api/src/main/java/org/dspace/uclouvain/consumer/FillerValueRemoverConsumer.java]: remove

File: dspace-api/src/main/java/org/dspace/uclouvain/consumer/FillerValueRemoverConsumer.java [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+                if (!mvsToClear.isEmpty()) {
+                    try {
+                        itemService.removeMetadataValues(
+                            context,
+                            item,
+                            mvsToClear
+                        );
+                        metadataModified = true;
```
Comment (zannkukai) [inline dspace-api/src/main/java/org/dspace/uclouvain/consumer/FillerValueRemoverConsumer.java]: remove

File: dspace-api/src/main/java/org/dspace/uclouvain/consumer/FillerValueRemoverConsumer.java [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+                            mv.getSchema(),
+                            mv.getElement(),
+                            mv.getQualifier(),
+                            mv.getLanguage(),
+                            CrisConstants.PLACEHOLDER_PARENT_METADATA_VALUE,
+                            mv.getAuthority(),
+                            mv.getConfidence(),
+                            mv.getPlace()
+                        );
+                        metadataModified = true;
```
Comment (zannkukai) [inline dspace-api/src/main/java/org/dspace/uclouvain/consumer/FillerValueRemoverConsumer.java]: remove

File: dspace-api/src/main/java/org/dspace/uclouvain/consumer/FillerValueRemoverConsumer.java [hunk 9]
```diff
@@ -0,0 +1,248 @@
...
+                    } catch (SQLException e) {
+                        logger.warn(
+                            "Could not replace with placeholder for metadata "
+                            + mv.getMetadataField() + " of item: " + uuid,
+                            e
+                        );
+                    }
+                }
+
+                if (metadataModified) {
```
Comment (zannkukai) [inline dspace-api/src/main/java/org/dspace/uclouvain/consumer/FillerValueRemoverConsumer.java]: `if (!(mvsToClear.isEmpty() && mvsToReplace.isEmpty())) `

File: dspace-api/src/main/java/org/dspace/uclouvain/consumer/FillerValueRemoverConsumer.java [hunk 10]
```diff
@@ -0,0 +1,248 @@
...
+        SubmissionStepConfig stepConfig,
+        DCInputSet inputSet,
+        List<String> groupFields,
+        DCInputsReader dciReader,
+        boolean isFormGroup
+    ) throws DCInputsReaderException {
+        for (DCInput[] fields: inputSet.getFields()) {
+            for (DCInput input: fields) {
+                // If the input has a 'group-like' type, search for 'children' fields and process them.
+                if (StringUtils.equalsAny(input.getInputType().toLowerCase(),
```
Comment (zannkukai) [inline dspace-api/src/main/java/org/dspace/uclouvain/consumer/FillerValueRemoverConsumer.java]: `input.getInputType()` could return null.... strange but it's true. So you need to test '!=null' before using 'tolowercase`

File: dspace-api/src/main/java/org/dspace/uclouvain/consumer/FillerValueRemoverConsumer.java [hunk 11]
```diff
@@ -0,0 +1,248 @@
...
+                    "group", "inline-group", "inline-labeled-group")) {
+                    // Get the form corresponding to the group field.
+                    DCInputSet inputSetGroup = dciReader.getInputsByFormName(stepConfig.getId() + "-"
+                        + Utils.standardize(input.getSchema(), input.getElement(), input.getQualifier(), "-"));
+                    // Extract the fields in the map by putting the 'inGroup' flag to true.
+                    extractFormGroupFields(stepConfig, inputSetGroup, groupFields, dciReader, true);
+                    // Skip below code cause we dont want to process the current input if it is of type group.
```
Comment (zannkukai) [inline dspace-api/src/main/java/org/dspace/uclouvain/consumer/FillerValueRemoverConsumer.java]: typo : "**be**cause you don**&pos;**t"

**Review by zannkukai — APPROVED** [review]
